core/src/avm2/object/stage_object.rs

```diff
@@ -29,40 +46,83 @@ pub struct StageObjectData<'gc> {
+        let proto = constr
+            .get_property(
+                constr,
+                &QName::new(Namespace::public(), "prototype"),
+                activation,
+            )?
+            .coerce_to_object(activation)?;
+
+        let mut instance = Self(GcCell::allocate(
+            activation.context.gc_context,
             StageObjectData {
-                base: ScriptObjectData::base_new(Some(proto), ScriptObjectClass::NoClass),
+                base: ScriptObjectData::base_new(
+                    Some(proto),
+                    ScriptObjectClass::ClassInstance(constr),
+                ),
                 display_object: Some(display_object),
             },
-        ))
+        ));
+        instance.install_instance_traits(activation, constr)?;
+
+        Ok(instance)
     }
 
-    /// Construct a stage object subclass.
-    pub fn derive(
-        base_proto: Object<'gc>,
-        mc: MutationContext<'gc, '_>,
-        class: GcCell<'gc, Class<'gc>>,
-        scope: Option<GcCell<'gc, Scope<'gc>>>,
-    ) -> Result<Object<'gc>, Error> {
-        let base = ScriptObjectData::base_new(
-            Some(base_proto),
-            ScriptObjectClass::InstancePrototype(class, scope),
-        );
+    /// Allocate and construct the AVM2 side of a display object intended to be
+    /// of a given constructor's type.
+    ///
+    /// This function is intended for display objects that do not have children
+    /// and thus do not need to be allocated and initialized in separate phases.
+    pub fn for_display_object_childless(
+        activation: &mut Activation<'_, 'gc, '_>,
+        display_object: DisplayObject<'gc>,
+        constr: Object<'gc>,
+    ) -> Result<Self, Error> {
+        let this = Self::for_display_object(activation, display_object, constr)?;
 
-        Ok(StageObject(GcCell::allocate(
-            mc,
+        constr.call_native_initializer(Some(this.into()), &[], activation, Some(constr))?;
+
+        Ok(this)
+    }
+
+    /// Create a `graphics` object for a given display object.
+    pub fn graphics_of(
+        activation: &mut Activation<'_, 'gc, '_>,
+        display_object: DisplayObject<'gc>,
+    ) -> Result<Self, Error> {
+        let constr = activation.avm2().constructors().graphics;
+        let proto = activation.avm2().prototypes().graphics;
+        let mut this = Self(GcCell::allocate(
+            activation.context.gc_context,
             StageObjectData {
-                base,
-                display_object: None,
+                base: ScriptObjectData::base_new(
+                    Some(proto),
+                    ScriptObjectClass::ClassInstance(constr),
+                ),
+                display_object: Some(display_object),
             },
-        ))
-        .into())
+        ));
+        this.install_instance_traits(activation, constr)?;
+
+        constr.call_native_initializer(Some(this.into()), &[], activation, Some(constr))?;
+
+        Ok(this)
```

Comment:
We need `graphics` to know *which* `DisplayObject` it's writing to, so we need the extra slot that `StageObject` has.

The only weirdness I think this enables is that you might be able to `mc.addChild(mc2.graphics)`, which would be equivalent to `mc.addChild(mc2)`. This is purely because our `addChild` doesn't have param metadata to enable the typecheck for `DisplayObject`.

Alternatively if this is too weird, we could have `Graphics` store it's associated display object on itself in the Ruffle namespace. It'd add an extra property lookup to graphics operations which is why I didn't want to do that.

Reply:
Maybe in the future if we have a slightly less-annoying way to do inheritance, we could make a separate `GraphicsObject` type that can contain its own pointer.